### PR TITLE
Libspark fuzzing harness

### PR DESCRIPTION
**Pull request by man2706kum**
This PR includes fuzzing harness for libspark. The harness is inside src/fuzz/. The quick start guide can be found in src/fuzz/README.md

### CODE REVIEW

File: src/fuzz/fuzzing_utilities.cpp [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        result[i].randomize();
+    }
+    return result;
+}
+
+std::vector<secp_primitives::GroupElement> FuzzedSecp256k1Object::GetGroupElements(int len) {
+    std::vector<secp_primitives::GroupElement> ge_vec;
+    ge_vec.resize(len);
+    for (int i = 0; i <= len; i++) {
+        ge_vec.push_back(GetGroupElement());
```
Comment (levonpetrosyan93) [inline src/fuzz/fuzzing_utilities.cpp:50]: resize creates len number of infinity GroupElements and pushes into ge_vec, and here we push len+1 number of GroupElements, is this function suppose to work that way?

File: src/fuzz/fuzzing_utilities.cpp [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    secp_primitives::GroupElement ge;
+    ge.randomize();
+    return ge;
+}
+
+std::vector<secp_primitives::GroupElement> FuzzedSecp256k1Object::GetMemberGroupElements(size_t len) {
+    std::vector<secp_primitives::GroupElement> ge_vec;
+    ge_vec.resize(len);
+    for (size_t i = 0; i <= len; i++) {
+        ge_vec[i] = (GetMemberGroupElement());
```
Comment (levonpetrosyan93) [inline src/fuzz/fuzzing_utilities.cpp:32]: i <= len will make loop go out of ge_vec's bound

File: src/fuzz/fuzzing_utilities.cpp [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    return ge_vec;
+}
+
+std::vector<secp_primitives::Scalar> FuzzedSecp256k1Object::GetScalars(size_t len) {
+    std::vector<secp_primitives::Scalar> scalar_vec;
+    scalar_vec.resize(len);
+    for (int i = 0; i <= len; i++) {
+        scalar_vec.push_back(GetScalar());
```
Comment (levonpetrosyan93) [inline src/fuzz/fuzzing_utilities.cpp:60]: the same here

File: src/fuzz/fuzzing_utilities.cpp [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    return scalar_vec;
+}
+
+std::vector<secp_primitives::Scalar> FuzzedSecp256k1Object::GetScalarsVector(size_t len) {
+    std::vector<secp_primitives::Scalar> scalar_vec;
+    scalar_vec.resize(len);
+    for (int i = 0; i <= len; i++) {
+        scalar_vec.push_back(GetScalar());
```
Comment (levonpetrosyan93) [inline src/fuzz/fuzzing_utilities.cpp:70]: same here

File: src/fuzz/fuzzing_utilities.cpp [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+secp_primitives::Scalar FuzzedSecp256k1Object::GetScalar_modified() {
+    secp_primitives::Scalar s = secp_primitives::Scalar(this->fdp->ConsumeBytes<uint8_t>(256).data());
+    return s;
+}
+
+std::vector<secp_primitives::Scalar> FuzzedSecp256k1Object::GetScalars_modified(int len) {
+    std::vector<secp_primitives::Scalar> scalar_vec;
+    scalar_vec.resize(len);
+    for (int i = 0; i <= len; i++) {
+        scalar_vec.push_back(GetScalar_modified());
```
Comment (levonpetrosyan93) [inline src/fuzz/fuzzing_utilities.cpp:85]: same here

File: src/fuzz/libspark/aead_debug
Comment (levonpetrosyan93) [inline src/fuzz/libspark/aead_debug:1]: please remove binary files

File: src/fuzz/libspark/aead_hfuzz
Comment (psolstice) [inline src/fuzz/libspark/aead_hfuzz:1]: Please remove binary files. I assume these are built from the sources

**Review by psolstice — CHANGES_REQUESTED** [review]

File: src/fuzz/fuzzing_utilities.cpp [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    for (size_t i = 0; i < len; i++) {
+        result[i].randomize();
+    }
+    return result;
+}
+
+std::vector<secp_primitives::GroupElement> FuzzedSecp256k1Object::GetGroupElements(int len) {
+    std::vector<secp_primitives::GroupElement> ge_vec;
+    ge_vec.resize(len);
+    for (int i = 0; i < len; i++) {
```
Comment (levonpetrosyan93) [inline src/fuzz/fuzzing_utilities.cpp:49]: We don't need this loop, as ge_vec.resize(len) does the same, or do reserve() instead of resize().

**Review by levonpetrosyan93 — CHANGES_REQUESTED** [review]

File: src/fuzz/README.md [hunk 7]
```diff
@@ -0,0 +1,119 @@
+# Fuzzing libspark
+
+## Quickstart Guide
+### Dependencies
```
Comment (AaronFeickert) [inline src/fuzz/README.md:4]: This is missing several `boost` and `openssl` dependencies that are required for the build to succeed.

File: src/fuzz/README.md [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+sudo apt-get install binutils-dev libunwind-dev libblocksruntime-dev clang
+git clone https://github.com/google/honggfuzz.git
+cd honggfuzz
+make
+sudo make install
+```
+
+2. Build firo: 
+
+Follow the instructions from https://github.com/firoorg/firo/tree/spark#readme
```
Comment (AaronFeickert) [inline src/fuzz/README.md:16]: This contradicts the `git clone` command in the next step.

File: src/fuzz/README.md [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+
+### Fuzzing using honggfuzz
+* In order to fuzz `firo/src/libpark` using Honggfuzz:
+
+```
+git clone -b spark_fuzz_blog https://github.com/hashcloak/firo.git
```
Comment (AaronFeickert) [inline src/fuzz/README.md:22]: This should be changed to the upstream repository.

File: src/fuzz/README.md [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+```
+The above command will generate an instrumented binary with name `<filename>_hfuzz` (eg: bpplus_hfuzz) inside src/fuzz/libspark.
+The fuzzing harness of the following spark files is availabe: aead, bech32, bpplus, chaum, coin, f4grumble, grootle, mint_transaction, schnorr and spend_transaction.
+
+* To start fuzzing:
+
+1. create a directory to save all the crahses.
+2. Inside the directory run:
+```
+hongfuzz -i <path_of_input_corpora>/<filename_inputs> -- ./<filename_hfuzz> ___FILE___
```
Comment (AaronFeickert) [inline src/fuzz/README.md:48]: `s/hongfuzz/honggfuzz`

File: src/fuzz/README.md [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+```
+
+example: 
+1. `mkdir src/fuzz/bpplus_results && cd src/fuzz/bpplus_results`
+2. `hongfuzz -i ../../inputs/bpplus_inputs -- ./../../libspark/bpplus_hfuzz ___FILE___`
```
Comment (AaronFeickert) [inline src/fuzz/README.md:53]: As above, `hongfuzz` is a typo. Also, the relative path to the instrumented binary is incorrect.

File: src/fuzz/README.md [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+3. To stop press `ctrl+c`
+
+If there is no input corpora, empty corpora can be provided.
```
Comment (AaronFeickert) [inline src/fuzz/README.md:56]: Please provide a note that this means creating and passing an empty directory via `-i`.

**Review by AaronFeickert — CHANGES_REQUESTED** [review]

**Review by levonpetrosyan93 — APPROVED** [review]

**Pull request merged**